cpp/src/arrow/util/stl_util_test.cc

```diff
@@ -55,5 +59,38 @@ TEST(StlUtilTest, VectorAddRemoveTest) {
   EXPECT_TRUE(result3.empty());
 }
 
+void ExpectSortPermutation(std::vector<std::string> unsorted,
+                           std::vector<int64_t> expected_indices) {
+  auto actual_indices = ArgSort(unsorted);
+  EXPECT_THAT(actual_indices, ::testing::ContainerEq(expected_indices));
+
+  auto sorted = unsorted;
+  std::sort(sorted.begin(), sorted.end());
+
+  auto permuted = unsorted;
+  Permute(expected_indices, &permuted);
+
+  EXPECT_THAT(permuted, ::testing::ContainerEq(sorted));
+}
+
+TEST(StlUtilTest, ArgSortPermute) {
+  std::string f = "foxtrot", a = "alpha", b = "bravo", d = "delta", c = "charlie",
+              e = "echo";
+
+  ExpectSortPermutation({a, f}, {0, 1});
+  ExpectSortPermutation({f, a}, {1, 0});
+  ExpectSortPermutation({a, b, c}, {0, 1, 2});
+  ExpectSortPermutation({a, c, b}, {0, 2, 1});
+  ExpectSortPermutation({c, a, b}, {1, 2, 0});
+  ExpectSortPermutation({a, b, c, d, e, f}, {0, 1, 2, 3, 4, 5});
+  ExpectSortPermutation({f, e, d, c, b, a}, {5, 4, 3, 2, 1, 0});
+  ExpectSortPermutation({f, a, b, d, c, e}, {1, 2, 4, 3, 5, 0});
+  ExpectSortPermutation({d, f, e, c, b, a}, {5, 4, 3, 0, 2, 1});
+  ExpectSortPermutation({b, a, c, d, f, e}, {1, 0, 2, 3, 5, 4});
+  ExpectSortPermutation({c, b, a, d, e, f}, {2, 1, 0, 3, 4, 5});
+  ExpectSortPermutation({b, c, a, f, d, e}, {2, 0, 1, 4, 5, 3});
+  ExpectSortPermutation({b, c, d, e, a, f}, {4, 0, 1, 2, 3, 5});
```

Comment:
This is non blocking.

Reply:
Since ArgSort is not stable the ordering of the indices of equivalent entries is not guaranteed. We could change ArgSort or add an option to make it stable